django/utils/functional.py

```diff
@@ -18,13 +21,30 @@ class cached_property:
     Decorator that converts a method with a single self argument into a
     property cached on the instance.
 
-    Optional ``name`` argument allows you to make cached properties of other
-    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
+    It is possible to make a cached property out of an existing method:
+    (e.g.  ``url = cached_property(get_absolute_url)`` ).
+    On Python versions before 3.6, you'll need to provide the optional ``name``
+    argument. (e.g.  ``url = cached_property(get_absolute_url, name='url')`` ).
     """
     def __init__(self, func, name=None):
+        func_name = func.__name__
+        unsupported_mangled_method = (
+            not PY36 and not name and
+            func_name.startswith('__') and not func_name.endswith('__')
+        )
+        if unsupported_mangled_method:
+            warnings.warn(
```

Comment:
Why the warning, but not an exception?

Reply:
Semver

Reply:
I also still don't understand why it's useful to allow writing code that doesn't work.

Reply:
well it does work, it just behaves as though `@cached_property` were not there.

Reply:
I think silently failing to cache the property should be considered as not working.

Reply:
I would consider that as not working

Reply:
What should the exception be?

Reply:
`ValueError` maybe since passing along `name` can fix it?